### ISSUE TITLE
glibc and libpng vulnerabilities

### ISSUE THREAD

**Issue by techie2012** [GoogleContainerTools/distroless#494]
gcr.io/distroless/java:8 has following vulnerabilities ,how to address them ?

CVE-2018-1000001
CVE-2018-20796
CVE-2019-9192
CVE-2019-6488
CVE-2017-12652
CVE-2017-9614

**Comment by yashbhutwala** [comment]
can you be more specific to the scanner used, the BOM generated, which components did the scanner say the CVEs come from?

**Comment by techie2012** [comment]
below are the details.
glibc 2.24-11+deb9u4
CVE-2018-1000001 - In glibc 2.26 and earlier there is confusion in the usage of getcwd() by realpath() which can be used to write before the destination buffer leading to a buffer underflow and potential code execution.
Show layers - bazel build

CVE-2018-20796 - In the GNU C Library (aka glibc or libc6) through 2.29, check_dst_limits_calc_pos_1 in posix/regexec.c has Uncontrolled Recursion, as demonstrated by '(\227|)(\\1\\1|t1|\\\2537)+' in grep.
- bazel build

CVE-2019-9192 - ** DISPUTED ** In the GNU C Library (aka glibc or libc6) through 2.29, check_dst_limits_calc_pos_1 in posix/regexec.c has Uncontrolled Recursion, as demonstrated by '(|)(\\1\\1)*' in grep, a different issue than CVE-2018-20796. NOTE: the software maintainer disputes that this is a vulnerability because the behavior occurs only with a crafted pattern.
Show layers  - bazel build

CVE-2019-6488- The string component in the GNU C Library (aka glibc or libc6) through 2.28, when running on the x32 architecture, incorrectly attempts to use a 64-bit register for size_t in assembly codes, which can lead to a segmentation fault or possibly unspecified other impact, as demonstrated by a crash in __memmove_avx_unaligned_erms in sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S during a memcpy - baxzel build

CVE-2019-7309 - the GNU C Library (aka glibc or libc6) through 2.29, the memcmp function for the x32 architecture can incorrectly return zero (indicating that the inputs are equal) because the RDX most significant bit is mishandled. - bazel buold


libpng 1.6.28-1+deb9u1

CVE-2017-12652 - libpng before 1.6.32 does not properly check the length of chunks against the user limit - bazel build

**Comment by chanseokoh** [comment]
I've checked latest Debian package snapshots for Debian 9 (stretch), and looks like Debian no longer provides higher versions of glibc and libpng on stretch. I think it's better to switch to use Distroless images based on Debian 10 (buster).

For Distroless Java, it means the only option is gcr.io/distroless/java-debian10, which runs `openjdk-11-jre-headless`. Unfortunately, Debian has dropped Java 8 on buster. But the Java 11 runtime should generally have no issue running Java 8 bytecode.

**Comment by techie2012** [comment]
this version of glibc has following vulnerabilities ,
glibc
Version: 2.28-10

CVE-2019-9192
In the GNU C Library (aka glibc or libc6) through 2.29, check_dst_limits_calc_pos_1 in posix/regexec.c has Uncontrolled Recursion, as demonstrated by '(|)(\\1\\1)*' in grep, a different issue than CVE-2018-20796. NOTE: the software maintainer disputes that this is a vulnerability because the behavior occurs only with a crafted pattern.

CVE-2018-20796
In the GNU C Library (aka glibc or libc6) through 2.29, check_dst_limits_calc_pos_1 in posix/regexec.c has Uncontrolled Recursion, as demonstrated by '(\227|)(\\1\\1|t1|\\\2537)+' in grep.

**Comment by chanseokoh** [comment]
[2.28-10](https://packages.debian.org/buster/libc6) is the latest that Debian provides. Hopefully, they roll out a newer version soon. I see they are preparing [2.30-4 in sid (unstable)](https://www.debian.org/releases/sid/).

**Comment by chanseokoh** [comment]
We implemented a Debian upstream package monitoring system (#544) about three weeks ago, and it has been working smoothly (automatically filing a PR to update packages such as #549). Distroless will now be able to closely follow the Debian upstream trajectory (Debian 10 buster and Debian 9 stretch, while I see they don't really update stretch packages). As of now, Distroless images are fully up-to-date. If you see CVEs now or going forward, it's a fair game; Debian buster and stretch would have the same CVEs (unless the monitoring system unexpected stops working properly). Closing the issue.
